csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs

```diff
@@ -282,6 +282,50 @@ public TrainingSession(SessionOptions options, CheckpointState state, string tra
             }
         }
 
+        /// <summary>
+        /// Computes the outputs of the training model and the gradients of the trainable parameters for the given OrtValue inputs
+        ///
+        /// This function performs a training step that computes the outputs of the training model and the gradients
+        /// of the trainable parameters for the given  OrtValue inputs. The train step is performed based on the training model
+        /// that was provided to the training session.
+        /// The TrainStep method is equivalent of running forward propagation and backward propagation in a single
+        /// step.
+        /// The gradients computed are stored inside the training session state so they can be later consumed
+        /// by the OptimizerStep function.
+        /// The gradients can be lazily reset by invoking the LazyResetGrad function.
+        /// Example usage:
+        /// <code>
+        /// var inputValues = new List&lt;OrtValue&gt;();
```

Comment:
The example leaks memory. It stores the input OrtValues inside a `List` that is not IDisposable. This makes it impossible to dispose them on exception at once. At the very least, the user would not bother to do so because it requires iteration of the list item by item.

Reply:
Thanks for the feedback. I have tried to address the leak in my new commit. Please let me know if that addresses your concern or not.

Reply:
This is a public interface, but `DisposableList` is an internal class that is not available for general use. I think it would be better to make it public and add a few tests. 

This would make user experience much better, and we could get rid of its copy in test code.

Reply:
My suggestion is to ore-prepend input ort values with `using` directive so they are disposed regardless of the List they are referred to.

Reply:
Thanks for all the help @yuslepukhin. I just submitted a commit with an update.